sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/EmptyCollectionsShouldNotBeEnumerated.cs

```diff
@@ -0,0 +1,306 @@
+                    {
+                        newProgramState = collectionSymbol.SetConstraint(CollectionConstraint.NotEmpty, newProgramState);
+                    }
+                    else
+                    {
+                        OnCollectionAccessed(instruction, collectionSymbol.HasConstraint(CollectionConstraint.Empty, programState));
+                    }
+                }
+
+                return newProgramState;
+            }
+
+            private ProgramState ProcessInvocation(ProgramState programState, InvocationExpressionSyntax instruction)
+            {
+                var newProgramState = programState;
+
+                var memberAccess = instruction.Expression as MemberAccessExpressionSyntax;
+                if (memberAccess != null)
+                {
+                    var collectionSymbol = semanticModel.GetSymbolInfo(memberAccess.Expression).Symbol;
+                    var collectionType = GetCollectionType(collectionSymbol);
+
+                    if (collectionType.IsAny(CollectionTypes))
+                    {
+                        var methodSymbol = semanticModel.GetSymbolInfo(instruction).Symbol as IMethodSymbol;
+                        if (methodSymbol == null ||
+                            methodSymbol.IsExtensionMethod ||
+                            IgnoredMethods.Contains(methodSymbol.Name))
+                        {
+                            return newProgramState;
+                        }
+
+                        if (AddMethods.Contains(methodSymbol.Name))
+                        {
+                            newProgramState = collectionSymbol.SetConstraint(CollectionConstraint.NotEmpty, newProgramState);
+                        }
+                        else
+                        {
+                            OnCollectionAccessed(instruction, collectionSymbol.HasConstraint(CollectionConstraint.Empty, programState));
+                        }
+                    }
+                }
+
+                newProgramState = RemoveCollectionConstraintsFromArguments(instruction, newProgramState);
+
+                return newProgramState;
+            }
+
+            private static ProgramState RemoveCollectionConstraintsFromArguments(InvocationExpressionSyntax instruction, ProgramState newProgramState)
+            {
+                var values = instruction.ArgumentList.Arguments.Select((node, index) =>
+                {
+                    SymbolicValue value;
+                    newProgramState.ExpressionStack.Pop(out value);
+                    return value;
+                });
+                foreach (var value in values)
+                {
+                    newProgramState = value.RemoveConstraint(CollectionConstraint.Empty, newProgramState);
+                }
+
+                return newProgramState;
+            }
+
+            public override ProgramState ObjectCreated(ProgramState programState, SymbolicValue symbolicValue, SyntaxNode instruction)

```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this method to reduce its Cognitive Complexity from 21 to the 15 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS3776)